src/borrow_tracker/tree_borrows/wildcard.rs

```diff
@@ -0,0 +1,409 @@
+                if old_max_child_access == new_max_child_access {
+                    // child_access didnt change, so we dont need to propagate further upwards
+                    break;
+                }
+                prev_old_access = old_max_child_access;
+                prev = id;
+            }
+        }
+        //traverses up the tree to update max_foreign_access fields
+        while let Some((id, access_type)) = stack.pop() {
+            let node = nodes.get(id).unwrap();
+            let mut entry = wildcard_accesses.entry(id);
+            let access = entry.or_insert(Default::default());
+            // all items on the stack need this updated
+            access.max_foreign_access = access_type;
+            let exposed_as = node.exposed_as(perms.get(id).map(|p| p.permission()));
+
+            push_relevant_children(
+                &mut stack,
+                /* other factors */ exposed_as,
+                access_type,
+                old_access_type,
+                access.clone(),
+                node.children.iter().copied(),
+                nodes,
+                perms,
+                wildcard_accesses,
+            );
+        }
+        #[cfg(feature = "expensive-consistency-checks")]
+        Self::verify_consistency(id, nodes, perms, wildcard_accesses)
+    }
+    /// verifies that the access tracking state is consistent
+    ///
+    /// panics if invalid
+    #[cfg(feature = "expensive-consistency-checks")]
+    pub fn verify_consistency(
+        id: UniIndex,
+        nodes: &UniValMap<Node>,
+        perms: &UniValMap<LocationState>,
+        wildcard_accesses: &UniValMap<WildcardAccessTracking>,
+    ) {
+        // find root node
+        let mut root = id;
+        while let Some(parent) = nodes.get(root).and_then(|n| n.parent) {
+            root = parent;
+        }
+
+        // TODO valid if map is empty
+        if !wildcard_accesses.contains_idx(root) {
+            return;
+        }
+        let mut stack = vec![root];
+        while let Some(id) = stack.pop() {
+            let node = nodes.get(id).unwrap();
+            stack.extend(node.children.iter());
+
+            let access = wildcard_accesses.get(id).unwrap();
+
+            let expected_max_foreign_access = if let Some(parent) = node.parent {
+                let parent_node = nodes.get(parent).unwrap();
+                let parent_perm = perms.get(parent).map(LocationState::permission);
+                let parent_access = wildcard_accesses.get(parent).unwrap();
+
+                let max_other_children = parent_node
+                    .children
+                    .iter()
+                    .copied()
+                    .filter(|child| *child != id)
+                    .map(|child| {
+                        let node = nodes.get(child).unwrap();
+                        let perm = perms.get(child).map(LocationState::permission);
+                        let access = wildcard_accesses.get(child).unwrap();
```

Comment:
This uses the data we are checking (`wildcard_accesses`) to check said data... doesn't that risk that we miss some inconsistencies?

For instance, what if we have a tree where no node is exposed, but there's a node somewhere that says "one of my children is exposed" and there's exactly one child that says "my parent is exposed"? Would we catch that?